### PR TITLE
Cake Thief Solution

### PR DESCRIPTION
**Pull request by robertwpaul**
To test:

```
ruby cake_thief_test.rb
Run options: --seed 55993

# Running:

.........

Finished in 0.000917s, 9814.6129 runs/s, 9814.6129 assertions/s.

9 runs, 9 assertions, 0 failures, 0 errors, 0 skips
```

(This was made slightly easier by having watched @awflwafl do it first).

### CODE REVIEW

File: cake_thief/robertwpaul/cake_thief_test.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+require 'minitest/autorun'
+require './cake_thief'
+
+class CakeThiefTest < Minitest::Test
+
+  def test_zero_capacity
+    cakes = [Cake.new(7, 160), Cake.new(3, 90), Cake.new(2, 15)]
+    capacity = 0
+    assert_equal 0, max_duffel_bag_value(cakes, capacity)
```
Comment (awflwafl) [inline cake_thief/robertwpaul/cake_thief_test.rb:9]: Could the same test-case be set up using a simpler setup? For example, just one cake.

By including unnecessary test setup logic we are giving our computers a larger set of instructions for no change in the outcome.

File: cake_thief/robertwpaul/cake_thief_test.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    capacity = 20
+    assert_equal 555, max_duffel_bag_value(cakes, capacity)
+  end
+
+  def test_sorting
+    input = [Cake.new(7, 160), Cake.new(3, 90), Cake.new(2, 15)]
+    expected = [Cake.new(3, 90), Cake.new(7, 160), Cake.new(2, 15)]
+    assert_equal expected, sorted_cakes(input)
+  end
+end
```
Comment (awflwafl) [inline cake_thief/robertwpaul/cake_thief_test.rb:59]: Is something like the following test case included for both solutions? This is something the group discussed in the Code Co-Op Slack channel.
```rb
def test_case
  cakes = [Cake.new(3, 16), Cake.new(2, 15)]
  capacity = 3
  assert_equal 16, max_duffel_bag_value(cakes, capacity)
end
```

It would be interesting to see failing/passing boundary-condition tests for each solution.

File: cake_thief/robertwpaul/cake_thief.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+Cake = Struct.new(:weight, :value)
+
+def sorted_cakes(cakes)
+  cakes.sort { |a, b| b.value / b.weight <=> a.value / a.weight }
```
Comment (awflwafl) [inline cake_thief/robertwpaul/cake_thief.rb:4]: First time I've seen this operator, nice. Today I learned about the _spaceship_ ruby operator 👍

**Review by awflwafl — APPROVED** [review]

File: cake_thief/robertwpaul/cake_thief_test.rb (hunk 2 above)
Comment (robertwpaul) [reply cake_thief/robertwpaul/cake_thief_test.rb:59]: It's included for the recursive solution - I took it out of the `sort-by-value-ratio` solution because it failed and as a professional software developer the solution to a failing test is to remove the test.

I'd be interested to see what the bounds of the recursive solution are. I think there are none, except memory and stack limitations. It essentially builds a binary tree, where each pair of children represents the choice of whether to take the current cake or not. Moving down the tree is solving the problem for smaller and smaller bags, all the way down to: `return 0 if index == cakes.length || capacity.zero?` - for a bag of size 0, the solution is 0. So I imagine it gets big.

I _think_ this is a non-optimal, greedy [Dynamic Programming](https://en.wikipedia.org/wiki/Dynamic_programming) solution. It's not concerned with, say, using the fewest _number_ of cakes to get the right answer, just getting the right answer.

File: cake_thief/robertwpaul/cake_thief_test.rb (hunk 1 above)
Comment (robertwpaul) [reply cake_thief/robertwpaul/cake_thief_test.rb:9]: Yeah, I think that's fair. I normally don't worry too much about test data that I think makes no difference to the outcome, but it's probably a good habit to worry a bit about it. Comes in handy once you start building larger test suites, with potentially expensive-to-create test data.

File: cake_thief/robertwpaul/cake_thief_test.rb (hunk 2 above)
Comment (awflwafl) [reply cake_thief/robertwpaul/cake_thief_test.rb:59]: > as a professional software developer the solution to a failing test is to remove the test

image
